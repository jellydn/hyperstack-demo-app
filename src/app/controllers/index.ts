// naming and imports: name and import as you see fit
import Auth from './auth'
import { NotesController } from './notes'

const controllers = []
controllers.push(Auth)
controllers.push(NotesController)

export const AppController = {
  controllers,
}
