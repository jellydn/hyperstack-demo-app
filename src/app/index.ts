import type { Context } from 'hyperstack'
import { context } from 'hyperstack'
import type * as Models from './models'
import type * as Mailers from './mailers'

export const appContext = context as Context<typeof Models, typeof Mailers>
