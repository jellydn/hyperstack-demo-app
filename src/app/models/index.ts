export { Note } from './note'
export { User } from './user'
