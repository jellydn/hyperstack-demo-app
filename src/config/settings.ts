// general purpose settings, with no infrastructure loading, no assumptions. keep it simple.
import path from 'path'
export const root = path.join(__dirname, '..')
