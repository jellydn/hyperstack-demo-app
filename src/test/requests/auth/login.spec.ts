import { test } from '@hyperstackjs/testing'
import { root } from '../../../config/settings'
import { appContext } from '../../../app'

const {
  requests,
  matchers: { matchRequestWithSnapshot },
  serializers: { redactAndExpectMatch },
} = test(root)

const serializer = redactAndExpectMatch({
  'body.token': /^eyJh.+/,
})

describe('requests', () => {
  describe('/auth/login', () => {
    requests('should login', async (request, _app, _server) => {
      const { User } = appContext.models()
      await User.createWithPassword({
        username: 'evh@example.com',
        password: 'mypass-should-login',
        name: 'Eddie Van Halen',
      })
      await matchRequestWithSnapshot(
        200,
        request().post('/auth/login').send({
          username: 'evh@example.com',
          password: 'mypass-should-login',
        }),
        { serializer }
      )
    })
    requests(
      'should not login on bad credentials',
      async (request, _app, _server) => {
        const { User } = appContext.models()
        await User.createWithPassword({
          username: 'evh@example.com',
          password: 'mypass-should-login',
          name: 'Eddie Van Halen',
        })
        await matchRequestWithSnapshot(
          401,
          request().post('/auth/login').send({
            username: 'evh@example.com',
            password: 'mypass-is-wrong',
          })
        )
      }
    )
    requests('should validate', async (request, _app, _server) => {
      const { User } = appContext.models()
      await User.createWithPassword({
        username: 'evh@example.com',
        password: 'mypass-should-login',
        name: 'Eddie Van Halen',
      })
      await matchRequestWithSnapshot(
        401,
        request().post('/auth/login').send({
          username: 'no-such-user@example.com',
          password: 'mypass-should-login',
        })
      )
      await matchRequestWithSnapshot(
        400,
        request().post('/auth/login').send({
          username: 'this-is-not-email',
          password: 'mypass-should-login',
        })
      )
      await matchRequestWithSnapshot(
        400,
        request().post('/auth/login').send({
          username: 'evh@example.com',
          password: 'boo',
        })
      )
    })
  })
})
