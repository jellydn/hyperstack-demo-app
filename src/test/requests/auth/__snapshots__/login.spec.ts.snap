// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`requests /auth/login should login 1`] = `
{
  "body": {
    "token": "test-redacted",
  },
  "headers": {
    "connection": "close",
    "content-length": "252",
    "content-security-policy": "default-src 'self';base-uri 'self';block-all-mixed-content;font-src 'self' https: data:;form-action 'self';frame-ancestors 'self';img-src 'self' data:;object-src 'none';script-src 'self';script-src-attr 'none';style-src 'self' https: 'unsafe-inline';upgrade-insecure-requests",
    "content-type": "application/json; charset=utf-8",
    "cross-origin-embedder-policy": "require-corp",
    "cross-origin-opener-policy": "same-origin",
    "cross-origin-resource-policy": "same-origin",
    "expect-ct": "max-age=0",
    "origin-agent-cluster": "?1",
    "referrer-policy": "no-referrer",
    "strict-transport-security": "max-age=15552000; includeSubDomains",
    "vary": "Accept-Encoding",
    "x-content-type-options": "nosniff",
    "x-dns-prefetch-control": "off",
    "x-download-options": "noopen",
    "x-frame-options": "SAMEORIGIN",
    "x-permitted-cross-domain-policies": "none",
    "x-xss-protection": "0",
  },
  "status": 200,
}
`;

exports[`requests /auth/login should not login on bad credentials 1`] = `
{
  "body": {
    "error": "incorrect username or password",
  },
  "headers": {
    "connection": "close",
    "content-length": "42",
    "content-security-policy": "default-src 'self';base-uri 'self';block-all-mixed-content;font-src 'self' https: data:;form-action 'self';frame-ancestors 'self';img-src 'self' data:;object-src 'none';script-src 'self';script-src-attr 'none';style-src 'self' https: 'unsafe-inline';upgrade-insecure-requests",
    "content-type": "application/json; charset=utf-8",
    "cross-origin-embedder-policy": "require-corp",
    "cross-origin-opener-policy": "same-origin",
    "cross-origin-resource-policy": "same-origin",
    "expect-ct": "max-age=0",
    "origin-agent-cluster": "?1",
    "referrer-policy": "no-referrer",
    "strict-transport-security": "max-age=15552000; includeSubDomains",
    "vary": "Accept-Encoding",
    "www-authenticate": "",
    "x-content-type-options": "nosniff",
    "x-dns-prefetch-control": "off",
    "x-download-options": "noopen",
    "x-frame-options": "SAMEORIGIN",
    "x-permitted-cross-domain-policies": "none",
    "x-xss-protection": "0",
  },
  "status": 401,
}
`;

exports[`requests /auth/login should validate 1`] = `
{
  "body": {
    "error": "incorrect username or password",
  },
  "headers": {
    "connection": "close",
    "content-length": "42",
    "content-security-policy": "default-src 'self';base-uri 'self';block-all-mixed-content;font-src 'self' https: data:;form-action 'self';frame-ancestors 'self';img-src 'self' data:;object-src 'none';script-src 'self';script-src-attr 'none';style-src 'self' https: 'unsafe-inline';upgrade-insecure-requests",
    "content-type": "application/json; charset=utf-8",
    "cross-origin-embedder-policy": "require-corp",
    "cross-origin-opener-policy": "same-origin",
    "cross-origin-resource-policy": "same-origin",
    "expect-ct": "max-age=0",
    "origin-agent-cluster": "?1",
    "referrer-policy": "no-referrer",
    "strict-transport-security": "max-age=15552000; includeSubDomains",
    "vary": "Accept-Encoding",
    "www-authenticate": "",
    "x-content-type-options": "nosniff",
    "x-dns-prefetch-control": "off",
    "x-download-options": "noopen",
    "x-frame-options": "SAMEORIGIN",
    "x-permitted-cross-domain-policies": "none",
    "x-xss-protection": "0",
  },
  "status": 401,
}
`;

exports[`requests /auth/login should validate 2`] = `
{
  "body": {
    "error": "validation",
    "validationErrors": [
      {
        "code": "invalid_string",
        "message": "Invalid email",
        "path": [
          "username",
        ],
        "validation": "email",
      },
    ],
  },
  "headers": {
    "connection": "close",
    "content-length": "136",
    "content-security-policy": "default-src 'self';base-uri 'self';block-all-mixed-content;font-src 'self' https: data:;form-action 'self';frame-ancestors 'self';img-src 'self' data:;object-src 'none';script-src 'self';script-src-attr 'none';style-src 'self' https: 'unsafe-inline';upgrade-insecure-requests",
    "content-type": "application/json; charset=utf-8",
    "cross-origin-embedder-policy": "require-corp",
    "cross-origin-opener-policy": "same-origin",
    "cross-origin-resource-policy": "same-origin",
    "expect-ct": "max-age=0",
    "origin-agent-cluster": "?1",
    "referrer-policy": "no-referrer",
    "strict-transport-security": "max-age=15552000; includeSubDomains",
    "vary": "Accept-Encoding",
    "x-content-type-options": "nosniff",
    "x-dns-prefetch-control": "off",
    "x-download-options": "noopen",
    "x-frame-options": "SAMEORIGIN",
    "x-permitted-cross-domain-policies": "none",
    "x-xss-protection": "0",
  },
  "status": 400,
}
`;

exports[`requests /auth/login should validate 3`] = `
{
  "body": {
    "error": "validation",
    "validationErrors": [
      {
        "code": "too_small",
        "inclusive": true,
        "message": "String must contain at least 5 character(s)",
        "minimum": 5,
        "path": [
          "password",
        ],
        "type": "string",
      },
    ],
  },
  "headers": {
    "connection": "close",
    "content-length": "185",
    "content-security-policy": "default-src 'self';base-uri 'self';block-all-mixed-content;font-src 'self' https: data:;form-action 'self';frame-ancestors 'self';img-src 'self' data:;object-src 'none';script-src 'self';script-src-attr 'none';style-src 'self' https: 'unsafe-inline';upgrade-insecure-requests",
    "content-type": "application/json; charset=utf-8",
    "cross-origin-embedder-policy": "require-corp",
    "cross-origin-opener-policy": "same-origin",
    "cross-origin-resource-policy": "same-origin",
    "expect-ct": "max-age=0",
    "origin-agent-cluster": "?1",
    "referrer-policy": "no-referrer",
    "strict-transport-security": "max-age=15552000; includeSubDomains",
    "vary": "Accept-Encoding",
    "x-content-type-options": "nosniff",
    "x-dns-prefetch-control": "off",
    "x-download-options": "noopen",
    "x-frame-options": "SAMEORIGIN",
    "x-permitted-cross-domain-policies": "none",
    "x-xss-protection": "0",
  },
  "status": 400,
}
`;
